libraries/lib-track/PendingTracks.cpp

```diff
@@ -0,0 +1,270 @@
+#include "PendingTracks.h"
+#include "Project.h"
+#include "Track.h"
+
+static const AudacityProject::AttachedObjects::RegisteredFactory
+sPendingTracksKey{
+  [](AudacityProject &project){
+     return std::make_shared<PendingTracks>(project);
+   }
+};
+
+PendingTracks &PendingTracks::Get(AudacityProject &project)
+{
+   return project.AttachedObjects::Get<PendingTracks>(sPendingTracksKey);
+}
+
+const PendingTracks &PendingTracks::Get(const AudacityProject &project)
+{
+   return Get(const_cast<AudacityProject &>(project));
+}
+
+PendingTracks::PendingTracks(AudacityProject &project)
+   : mTracks{ TrackList::Get(project) }
+   , mTrackListSubscription { mTracks.Subscribe(
+      [this](const TrackListEvent &event){
+         switch (event.mType) {
+         case TrackListEvent::PERMUTED:
+         case TrackListEvent::RESIZING:
+         case TrackListEvent::ADDITION:
+         case TrackListEvent::DELETION:
+            UpdatePendingTracks();
+            break;
+         default:
+            break;
+         }
+         // Pass along to downstream listeners
+         Publish(event);
+   })}
+   , mPendingUpdates{ TrackList::Temporary(mTracks.GetOwner()) }
+{}
+
+PendingTracks::~PendingTracks() = default;
+
+void PendingTracks::RegisterPendingNewTracks(TrackList &&list)
+{
+   mTracks.Append(std::move(list), false);
+}
+
+namespace {
+// function-making function
+auto finder(TrackId id, int &distance) {
+   // Predicate returns true if any channel if the track's channel group has
+   // the given id and as a side-effect reports the position of the channel in
+   // the group
+   return [id, &distance](const auto &pTrack) {
+      const auto channels = TrackList::Channels(&*pTrack);
+      const auto begin = channels.begin();
+      const auto end = channels.end();
+      const auto pred = [id](const auto &ptr) { return ptr->GetId() == id; };
+      auto iter = std::find_if(begin, end, pred);
+      if (iter == end) {
+         distance = -1;
+         return false;
+      }
+      else {
+         distance = std::distance(begin, iter);
+         return true;
+      }
+   };
+}
+}
+
+std::shared_ptr<Track>
+PendingTracks::SubstitutePendingChangedTrack(Track &track) const
+{
+   // Linear search.  Tracks in a project are usually very few.
+   auto pTrack = &track;
+   // track might not be a leader
+   if (!mPendingUpdates->empty()) {
+      const auto id = track.GetId();
+      const auto end = mPendingUpdates->end();
+      int distance{ -1 };
+      // Find the leader of the group of shadow tracks containing the id
+      if (const auto it =
+          std::find_if(mPendingUpdates->begin(), end, finder(id, distance))
+          ; it != end) {
+         // Find the correct corresponding channel
+         auto channelIter = TrackList::Channels(&**it).begin();
+         std::advance(channelIter, distance);
+         pTrack = *channelIter;
+      }
+   }
+   return pTrack->SharedPointer();
```

Comment:
This is a bit different from what `Track::SubstitutePendingChangedTrack` did. What if `track` isn't owned by the project's track list?

Reply:
Though `TrackList::sCounter` is `static`, so id's will unlikely collide